File: golangR2/solution/solution0106.go
package solution

func BuildTree(inorder []int, postorder []int) *TreeNode {
	var bf func(inorder []int, postorder []int) *TreeNode
	bf = func(inorder []int, postorder []int) *TreeNode {
		var root TreeNode
		if len(inorder) == 1 {
			return &TreeNode{Val: inorder[0]}
		}

		if len(inorder) == 2 {
			root = TreeNode{Val: postorder[1]}
			if root.Val == inorder[0] {
				r := TreeNode{Val: inorder[1]}
				root.Right = &r
			} else {
				l := TreeNode{Val: postorder[0]}
				root.Left = &l
			}
			return &root
		}

		if len(inorder) == 3 {
			root = TreeNode{Val: postorder[2]}
			l := TreeNode{Val: postorder[0]}
			r := TreeNode{Val: postorder[1]}
			root.Left = &l
			root.Right = &r
			return &root
		}

		val := postorder[len(postorder)-1]
		index := -1

		for i := 0; i < len(inorder); i++ {
			if inorder[i] == val {
				index = i
			}
		}

		root.Val = val
		inorder1 := inorder[0:index]
		inorder2 := inorder[index+1:]

		postorder1 := postorder[0:len(inorder1)]
		postorder2 := postorder[len(inorder1) : len(postorder)-1]

		root.Left = bf(inorder1, postorder1)
		root.Right = bf(inorder2, postorder2)

		return &root
	}

	return bf(inorder, postorder)
}

File: JavaR3/src/solution/Solution0106.java
package solution;

public class Solution0106 {
	//中序 后序
	public TreeNode buildTree(int[] inorder, int[] postorder) {
		TreeNode res;
		if (postorder.length == 1) {
			res = new TreeNode(postorder[0]);
		} else if (postorder.length == 0) {
			res = null;
		} else if (postorder.length == 2) {
			res = new TreeNode(postorder[1]);
			if (postorder[1] == inorder[0]) {
				res.left = null;
				res.right = new TreeNode(inorder[1]);
			} else {
				res.left = new TreeNode(inorder[0]);
				res.right = null;
			}
		} else {
			int postVal = postorder[postorder.length - 1];
			res = new TreeNode(postVal);
			int index = 0;
			for (int i = 0; i < inorder.length; i++) {
				if (inorder[i] == postVal) {
					index = i;
				}
			}

			int[] inOrderLeft = new int[index];
			int[] postOrderLeft = new int[index];

			System.arraycopy(inorder, 0, inOrderLeft, 0, index);
			System.arraycopy(postorder, 0, postOrderLeft, 0, index);

			int[] inorderRight = new int[inorder.length - index - 1];
			int[] postorderRight = new int[postorder.length - index - 1];

			System.arraycopy(inorder, index + 1, inorderRight, 0, inorder.length - index - 1);
			System.arraycopy(postorder, index, postorderRight, 0, postorder.length - index - 1);

			res.left = buildTree(inOrderLeft, postOrderLeft);
			res.right = buildTree(inorderRight, postorderRight);
		}

		return res;
	}
}
